### PR TITLE
A force-field compatible to smotion that mimics a time-dependent drive

### PR DESCRIPTION
**Pull request by mahrossi**
This has been tested and used by me and also by Karen at points in the past year. Everything is functional and there is an example added. The reference that gives more details to the implementation is cited. The physics of the problem itself could be improved and we are working on it, but as an infrastructure it is fine.

### CODE REVIEW

**Comment by mahrossi** [comment]
Now all tests pass. I will now add a more meaningful example for a solid as well, but that is using FHI-aims. If someone can review it would be great.

**Comment by mahrossi** [comment]
This needs a regtest, which can now exist because of the new merge.

**Comment by mahrossi** [comment]
Can this branch be reviewed? e.g. by @venkatkapil24 ? I am figuring out how to do the regtest but will deploy that tomorrow morning.

**Comment by mahrossi** [comment]
The regtest is added but I shall see if it works. It is really ready for review though (tagging again @venkatkapil24). I added one more example for a solid also.

One thing I am not sure how to do is how to make the functionality go directly to the latex documentation. Will try to figure out what @ceriottm did in the committee branch and follow the steps.

**Comment by venkatkapil24** [comment]
Sure, I will take a look. Currently it seems that the tests are not successful.

**Comment by mahrossi** [comment]
Yes, it is because of the new regtest. Could you or @litman90 help me, maybe? It is because it needs 2 sockets, one being a "native" i-PI forcefield. I tried to reproduce the steps but apparently failed.

**Comment by litman90** [comment]
> Yes, it is because of the new regtest. Could you or @litman90 help me, maybe? It is because it needs 2 sockets, one being a "native" i-PI forcefield. I tried to reproduce the steps but apparently failed.

sure, let me check it

**Comment by litman90** [comment]
For your reference, the main problem was with the naming of the files and the files_to_check.txt (for example it was *vel_1.out and *vel_2.out instead of vel_0.out and vel_1.out). In short, they were not matching.
I added some lines to the README in the reg_test file to explain it.

Please have a look at the small PR I merged into the sabia-group:dmdpy for more details

**Comment by mahrossi** [comment]
> For your reference, the main problem was with the naming of the files and the files_to_check.txt (for example it was *vel_1.out and *vel_2.out instead of vel_0.out and vel_1.out). In short, they were not matching.
> I added some lines to the README in the reg_test file to explain it.

No worries, this was just me working without paying attention -- that was clear hehe...

**Comment by mahrossi** [comment]
ping here -- reviews anyone?!

**Comment by mahrossi** [comment]
> One thing I am not sure how to do is how to make the functionality go directly to the latex documentation. Will try to figure out what @ceriottm did in the committee branch and follow the steps.

I managed the inclusion in the manual also now. So both regtests, examples, and manual entry are in. It is as ready as I can make it without a second pair of eyes going through it!

File: ipi/engine/forcefields.py
Comment (venkatkapil24) [inline ipi/engine/forcefields.py:454]: Can you provide a full citation ?  and describe the potential in a bit more detail ?

File: ipi/engine/smotion/dmd.py
Comment (venkatkapil24) [inline ipi/engine/smotion/dmd.py:18]: Description is not updated.

File: ipi/inputs/smotion/dmd.py
Comment (venkatkapil24) [inline ipi/inputs/smotion/dmd.py:1]: Can this be updated ?

**Comment by venkatkapil24** [comment]
Is there a special reason why the time dependent force is not simply added as a forcefield, but as a bias ?

File: ipi/engine/forcefields.py
Comment (venkatkapil24) [inline ipi/engine/forcefields.py:550]: Can you please also add a virial calculator, or have a way of raising an error ?

Comment (mahrossi) [reply ipi/engine/forcefields.py:550]: I'll  add the virial, it seems very easy for this specific ff. Thanks. However, should I also add a warning then for other current ff's that assign only 0 to the virials? (e.g. Debye and LJ)?

File: ipi/engine/smotion/dmd.py
Comment (mahrossi) [reply ipi/engine/smotion/dmd.py:18]: great, will do.

File: ipi/inputs/smotion/dmd.py
Comment (mahrossi) [reply ipi/inputs/smotion/dmd.py:1]: Yes, thanks!

File: ipi/engine/forcefields.py
Comment (mahrossi) [reply ipi/engine/forcefields.py]: Not yet ready but on the way.

**Comment by mahrossi** [comment]
> Is there a special reason why the time dependent force is not simply added as a forcefield, but as a bias ?

This is now explained in the docstrings.

File: ipi/engine/forcefields.py
Comment (mahrossi) [reply ipi/engine/forcefields.py]: ok

**Comment by mahrossi** [comment]
Hi this is done for now. I would like to learn about a standard way of benchmarking the virial term of the force fields. This seems pretty correct, but I like hard numbers.

File: examples/ffdmd/Si-drive/README
Comment (ceriottm) [inline examples/ffdmd/Si-drive/README:10]: what does this mean? shouldn't the script be included in the PR if it's needed to run the example?

File: examples/ffdmd/Si-drive/README
Comment (ceriottm) [inline examples/ffdmd/Si-drive/README:16]: Again, the script should be part of the PR, so this comment in the README should be changed

File: examples/ffdmd/Si-drive/README
Comment (ceriottm) [inline examples/ffdmd/Si-drive/README:1]: An explanation of what is the form of the time-dependent drive would help understand why you need all of those manipulations below

File: ipi/engine/forcefields.py
Comment (ceriottm) [inline ipi/engine/forcefields.py:452]: Why is this both a smotion class and a FF? It's awkward because now one can also use it as a forcefield outside an actual DMD calculation

File: ipi/engine/forcefields.py
Comment (ceriottm) [inline ipi/engine/forcefields.py:518]: You OK having this as the first version or you want to get more features in?

File: ipi/engine/smotion/dmd.py
Comment (ceriottm) [inline ipi/engine/smotion/dmd.py:56]: I realize this is very similar to how metadynamics is implemented, but it's fugly. Can we think of a better way?

File: ipi_tests/regression_tests/tests/DRIVEN/test_settings.dat
Comment (ceriottm) [inline ipi_tests/regression_tests/tests/DRIVEN/test_settings.dat:4]: Commenting here for the whole regtest - wouldn't it be better to call it "dmd" rather than "DRIVEN" so that the connection to the code is more apparent?

File: ipi/engine/smotion/dmd.py
Comment (mahrossi) [reply ipi/engine/smotion/dmd.py:56]: Now, I cannot.

File: ipi/engine/forcefields.py
Comment (mahrossi) [reply ipi/engine/forcefields.py:518]: No, I don't think this needs any new features, but still it is good to call attention to what is there and what is not. It uses just minimal im. conv.

File: ipi/engine/forcefields.py
Comment (mahrossi) [reply ipi/engine/forcefields.py:452]: I don't understand what you would like to see done here. The reason to do it like this is explained. You would like me to code the whole procedure as only the dmd part?

Comment (mahrossi) [reply ipi/engine/forcefields.py:452]: If someone wants to use it in some way out of the dmd step I don't see why not actually.

File: examples/ffdmd/Si-drive/README
Comment (mahrossi) [reply examples/ffdmd/Si-drive/README:1]: They are not needed and have been removed.

File: examples/ffdmd/Si-drive/README
Comment (mahrossi) [reply examples/ffdmd/Si-drive/README]: Not relevant anymore.

File: examples/ffdmd/Si-drive/README
Comment (mahrossi) [reply examples/ffdmd/Si-drive/README]: Also not relevant anymore.

**Comment by mahrossi** [comment]
Can this be merged now? I need an approval.

**Review by ceriottm — APPROVED** [review]

**Pull request merged**
